shortestBridge.js

```diff
@@ -0,0 +1,187 @@
+
+    return isRowBound && isColBound;
+};
+
+const getRowColString = ({ row, col }) => `${row}, ${col}`;
+
+const getChildren = (row, col, grid) => {
+    const children = [];
+
+    for (const direction of directions) {
+        const [rowDir, colDir] = direction;
+
+        const newRow = row + rowDir;
+        const newCol = col + colDir;
+
+        if (!isBound(newRow, newCol, grid)) continue;
+
+        children.push({ row: newRow, col: newCol });
+    }
+
+    return children;
+};
+
+const findFirstIslandCoordinates = (startRow, startCol, grid) => {
+    const visited = new Set();
+    const queue = new Queue();
+
+    const rowColString = getRowColString({ row: startRow, col: startCol });
+    visited.add(rowColString);
+
+    queue.enqueue({ row: startRow, col: startCol });
+
+    const firstIslandCoordinates = [];
+    while (queue.size() > 0) {
+        // Remove node
+        const { row, col } = queue.dequeue();
+
+        // Process node
+        if (grid[row][col] !== LAND) continue;
+        firstIslandCoordinates.push({ row, col });
+
+        // Get children
+        const children = getChildren(row, col, grid);
+        for (const child of children) {
+            const rowColString = getRowColString(child);
+            if (visited.has(rowColString)) continue;
+            visited.add(rowColString);
+
+            queue.enqueue({ ...child });
+        }
+    }
+
+    return firstIslandCoordinates;
+};
+
+const getFirstIslandCoordinates = (grid) => {
+    const numRows = grid.length;
+    const numCols = grid[0].length;
+
+    let firstIslandCoordinates = null;
+
+    for (let row = 0; row < numRows; row++) {
+        for (let col = 0; col < numCols; col++) {
+            if (grid[row][col] !== LAND) continue;
+
+            firstIslandCoordinates = findFirstIslandCoordinates(row, col, grid);
+
+            return firstIslandCoordinates;
+        }
+    }
+};
+
+const getShortestBridge = (grid, firstIslandCoordinates) => {
+    const visited = new Set();
+    const queue = new Queue();
+
+    for (const firstIslandCoordinate of firstIslandCoordinates) {
+        visited.add(getRowColString(firstIslandCoordinate));
+        queue.enqueue({ ...firstIslandCoordinate, path: 0 });
+    }
+
+    while (queue.size() > 0) {
+        // Remove node
+        const { row, col, path } = queue.dequeue();
+
+        // Process node
+        const isIslandVisited = visited.has(getRowColString({ row, col }));
+        const terrain = grid[row][col];
+
+        if (!isIslandVisited && terrain === LAND) {
+            const bridgeNum = path - 1;
+            return bridgeNum;
+        }
+
+        visited.add({ row, col });
+
+        // Get children
+        const children = getChildren(row, col, grid);
+        for (const child of children) {
+            const rowColString = getRowColString(child);
+            if (visited.has(rowColString)) continue;
+
+            queue.enqueue({ ...child, path: path + 1 });
+        }
+    }
```

Comment:
It would be best to have something explicitly handling the case where we never find a second island. I know it's not required by LeetCode, but it would be good to at least throw an error or something.

Reply:
Updated.

Reply:
I think it's significantly better to throw an error than just return a constant. In this case, the problem guarantees that there will be a second island. So, if we ever reach a situation where there is no second island, something has gone horribly wrong. In this case, we want to "fail loudly". If you're not familiar with that term, read this short blog post: http://oncodingstyle.blogspot.com/2008/10/fail-early-fail-loudly.html

Your current code would fail quietly because you'd just return a negative integer. It's very possible that the rest of the system will proceed as usual and try to use the -1. But, what you want is for the system to blow up immediately and show the developer that something went wrong.

Reply:
Interesting but good points. I've never heard about  "fail loudly" but it makes sense. I'll definitely checkout the article.

Reply:
Make sure to update the code! As of now, you're still failing quietly.

Reply:
I believe I got it correct after reading the blog. Let me know what you think.

Reply:
Great!